data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java

```diff
@@ -229,27 +293,68 @@ public void addValue(int rowId, byte[] data, ColumnVector output) {
         output.isNull[rowId] = true;
       } else {
         output.isNull[rowId] = false;
-        // getBinary always makes a copy, so we don't need to worry about it
-        // being changed behind our back.
         ((BytesColumnVector) output).setRef(rowId, data, 0, data.length);
       }
     }
   }
 
-  static class TimestampConverter implements Converter<Long> {
+  static class DateConverter implements Converter<LocalDate> {
     @Override
-    public Class<Long> getJavaClass() {
-      return Long.class;
+    public Class<LocalDate> getJavaClass() {
+      return LocalDate.class;
     }
 
-    public void addValue(int rowId, Long data, ColumnVector output) {
+    public void addValue(int rowId, LocalDate data, ColumnVector output) {
+      if (data == null) {
+        output.noNulls = false;
+        output.isNull[rowId] = true;
+      } else {
+        output.isNull[rowId] = false;
+        ((LongColumnVector) output).vector[rowId] = ChronoUnit.DAYS.between(EPOCH_DAY, data);
+      }
+    }
+  }
+
+  static class TimestampTzConverter implements Converter<OffsetDateTime> {
+    @Override
+    public Class<OffsetDateTime> getJavaClass() {
+      return OffsetDateTime.class;
+    }
+
+    public void addValue(int rowId, OffsetDateTime data, ColumnVector output) {
+      if (data == null) {
+        output.noNulls = false;
+        output.isNull[rowId] = true;
+      } else {
+        output.isNull[rowId] = false;
+        TimestampColumnVector cv = (TimestampColumnVector) output;
+        long micros = ChronoUnit.MICROS.between(EPOCH, data);
+        cv.time[rowId] = micros / 1_000; // millis
+        cv.nanos[rowId] = (int) (micros % 1_000_000) * 1_000; // nanos
```

Comment:
Doesn't this report milliseconds twice?

```
OffsetDateTime now = OffsetDateTime.now(); // 2020-03-17T00:21:46.844Z
long micros = ChronoUnit.MICROS.between(EPOCH, now); // 1584404506844000
long millis = micros / 1_000; // 1584404506844 <--- note 844 is millis
long nanos = (micros % 1_000_000) * 1_000; // 844000000 <--- millis are included
```

Does the millis/nanos format discard digits above 1,000,000 by convention? (ping @omalley)

Reply:
Yes we report the millis twice. ORC code only keeps seconds from cv.time and so handles this transparently. Since ORC expects the double reporting, I think it should be fine if we keep double reporting. @omalley Thoughts?

Note that the ORC reader double reports millis even if the writer does not double report millis.  Details in https://issues.apache.org/jira/browse/ORC-546

Reply:
Let's make sure this behavior is clear in the spec.

Reply:
Added this behaviour as a note for the timestamp types in the ORC type mapping in the spec